wardrobe/tradable_kitties/src/lib.rs

```diff
@@ -0,0 +1,389 @@
+    ListKittyForSale,
+    /// Delist the kitty from sale, This means tradable kitty will converted back to kitty.
+    DelistKittyFromSale,
+    /// Update the `price` of tradable kitty.
+    UpdateKittyPrice,
+    // Update the name of the kitty.
+    UpdateKittyName,
+    /// For buying a new kitty from other owners.
+    Buy,
+}
+
+/// Extract basic kitty data from a list of dynamically typed `TradableKitty` data, populating a list with basic kitty data.
+fn extract_basic_kitty_list(
+    tradable_kitty_data: &[DynamicallyTypedData],
+    kitty_data_list: &mut Vec<DynamicallyTypedData>,
+) -> Result<(), TradeableKittyError> {
+    for utxo in tradable_kitty_data {
+        if let Ok(tradable_kitty) = utxo.extract::<TradableKittyData>() {
+            kitty_data_list.push(tradable_kitty.kitty_basic_data.clone().into());
+        } else {
+            return Err(TradeableKittyError::BadlyTyped);
+        }
+    }
+    Ok(())
+}
+
+/// Checks if buying the kitty is possible or not. It depends on the Money variable to validate the spending of coins.
+/// Make sure to place the kitty first and then the coins in the transaction.
+fn check_can_buy<const ID: u8>(
+    input_data: &[DynamicallyTypedData],
+    output_data: &[DynamicallyTypedData],
+) -> Result<TransactionPriority, TradeableKittyError> {
+    let mut input_coin_data: Vec<DynamicallyTypedData> = Vec::new();
+    let mut output_coin_data: Vec<DynamicallyTypedData> = Vec::new();
+    let input_kitty_to_be_traded: Option<TradableKittyData>;
+
+    let mut total_input_amount: u128 = 0;
+    let total_price_of_kitty: u128;
+
+    if let Ok(td_input_kitty) = input_data[0].extract::<TradableKittyData>() {
+        ensure!(
+            td_input_kitty.price != 0,
+            TradeableKittyError::KittyPriceCantBeZero
+        );
+        input_kitty_to_be_traded = Some(td_input_kitty.clone());
+        total_price_of_kitty = td_input_kitty.price;
+    } else {
+        return Err(TradeableKittyError::BadlyTyped);
+    }
+
+    if let Ok(td_output_kitty) = output_data[0].extract::<TradableKittyData>() {
+        ensure!(
+            input_kitty_to_be_traded.clone().unwrap().kitty_basic_data
+                == td_output_kitty.kitty_basic_data,
+            TradeableKittyError::KittyBasicPropertiesAltered
+        );
+    } else {
+        return Err(TradeableKittyError::BadlyTyped);
+    }
+
+    for coin_data in input_data.iter().skip(1) {
+        let coin = coin_data
+            .clone()
+            .extract::<Coin<ID>>()
+            .map_err(|_| TradeableKittyError::BadlyTyped)?;
+
+        let utxo_value = coin.0;
+        ensure!(
+            utxo_value > 0,
+            TradeableKittyError::MoneyError(MoneyError::ZeroValueCoin)
+        );
+        input_coin_data.push(coin_data.clone());
+        total_input_amount = total_input_amount
+            .checked_add(utxo_value)
+            .ok_or(TradeableKittyError::MoneyError(MoneyError::ValueOverflow))?;
+    }
```

Comment:
This might be a little bias of mine but I like the big Rust one-liners instead of instantiating empty Vecs and pushing each time, which is also way less efficient.
Same below

Reply:
I prefer chaining iterator methods in most cases, but anything that passes clippy and isn't ridiculous is welcome here.